src/lib.rs

```diff
@@ -1,3 +1,162 @@
-#[test]
-fn it_works() {
+#[cfg(test)]
+mod test {
+    use tokens::{ Lexer, LexerOptions };
+
+    #[test]
+    fn lexer_usage() {
+        // build the lexer once for project environment with extensions.
+        let lexer = Lexer::new(LexerOptions::default(), vec![]);
+
+        // use many times.
+        {
+            let source: String = "{{ var }}".into();
+            for token in lexer.tokens(&source) {
```

Comment:
Ok, cool. We will have to check whether tail-calls are optimized in this case by LLVM. Tail-calls seem not to be guaranteed natively by rust. It will depend on the good-will of LLVM and possibly compiler flags.
https://doc.rust-lang.org/complement-design-faq.html#no-guaranteed-tail-call-optimization
https://mail.mozilla.org/pipermail/rust-dev/2013-April/003557.html